File: app/src/main/java/org/dicio/dicio_android/MainActivity.java
package org.dicio.dicio_android;

import static android.Manifest.permission.RECORD_AUDIO;
import static android.content.pm.PackageManager.PERMISSION_GRANTED;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.InputType;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.ScrollView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.core.app.ActivityCompat;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.preference.PreferenceManager;

import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton;
import com.google.android.material.navigation.NavigationView;

import org.dicio.dicio_android.eval.SkillEvaluator;
import org.dicio.dicio_android.eval.SkillRanker;
import org.dicio.dicio_android.input.InputDevice;
import org.dicio.dicio_android.input.SpeechInputDevice;
import org.dicio.dicio_android.input.ToolbarInputDevice;
import org.dicio.dicio_android.input.VoskInputDevice;
import org.dicio.dicio_android.output.graphical.MainScreenGraphicalDevice;
import org.dicio.dicio_android.output.speech.AndroidTtsSpeechDevice;
import org.dicio.dicio_android.output.speech.NothingSpeechDevice;
import org.dicio.dicio_android.output.speech.SnackbarSpeechDevice;
import org.dicio.dicio_android.output.speech.ToastSpeechDevice;
import org.dicio.dicio_android.settings.SettingsActivity;
import org.dicio.dicio_android.skills.SkillHandler;
import org.dicio.dicio_android.util.BaseActivity;
import org.dicio.dicio_android.util.PermissionUtils;
import org.dicio.skill.output.GraphicalOutputDevice;
import org.dicio.skill.output.SpeechOutputDevice;

public class MainActivity extends BaseActivity
        implements NavigationView.OnNavigationItemSelectedListener {

    public static final int MICROPHONE_PERMISSION_REQUEST_CODE = 13893;
    public static final int SKILL_PERMISSIONS_REQUEST_CODE = 1928430;
    public static final int SETTINGS_PERMISSIONS_REQUEST_CODE = 420938;

    private SharedPreferences preferences;

    private DrawerLayout drawer;
    private MenuItem textInputItem = null;

    @Nullable private SkillEvaluator skillEvaluator = null;
    private boolean appJustOpened = false;
    private boolean resumingFromSettings = false;
    private boolean textInputItemFocusJustChanged = false;


    ////////////////////////
    // Activity lifecycle //
    ////////////////////////

    @Override
    protected void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        preferences = PreferenceManager.getDefaultSharedPreferences(this);

        final Toolbar toolbar = findViewById(R.id.toolbar);
        drawer = findViewById(R.id.drawer_layout);
        final NavigationView navigationView = findViewById(R.id.nav_view);

        setSupportActionBar(toolbar);
        final ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.drawer_open, R.string.drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();
        navigationView.setNavigationItemSelectedListener(this);

        final ScrollView scrollView = findViewById(R.id.outputScrollView);
        scrollView.addOnLayoutChangeListener((v, left, top, right, bottom,
                                              oldLeft, oldTop, oldRight, oldBottom) -> {
            if (textInputItemFocusJustChanged && (oldBottom != bottom || oldTop != top)) {
                textInputItemFocusJustChanged = false; // the keyboard was opened because of menu
                scrollView.postDelayed(() ->
                        scrollView.scrollBy(0, oldBottom - bottom + top - oldTop), 10);
            }
        });

        appJustOpened = true; // determines whether to show initial panel and start listening
        initializeSkillEvaluator();
        setupVoiceButton();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        destroySkillEvaluator();
        SkillHandler.releaseSkillContext();
    }

    private void setupVoiceButton() {
        final ExtendedFloatingActionButton voiceFab = findViewById(R.id.voiceFab);
        final ProgressBar voiceLoading = findViewById(R.id.voiceLoading);
        if (skillEvaluator != null
                && skillEvaluator.getPrimaryInputDevice() instanceof SpeechInputDevice) {
            voiceFab.setVisibility(View.VISIBLE);
            voiceLoading.setVisibility(View.VISIBLE);
            ((SpeechInputDevice) skillEvaluator.getPrimaryInputDevice())
                    .setVoiceViews(voiceFab, voiceLoading);
        } else {
            voiceFab.setVisibility(View.GONE);
            voiceLoading.setVisibility(View.GONE);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(final Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);

        final ToolbarInputDevice toolbarInputDevice;
        if (skillEvaluator == null) {
            toolbarInputDevice = null;
        } else if (skillEvaluator.getPrimaryInputDevice() instanceof ToolbarInputDevice) {
            toolbarInputDevice = (ToolbarInputDevice) skillEvaluator.getPrimaryInputDevice();
        } else {
            toolbarInputDevice = skillEvaluator.getSecondaryInputDevice();
        }

        if (toolbarInputDevice != null) {
            textInputItem = menu.findItem(R.id.action_text_input);
            textInputItem.setVisible(true);

            toolbarInputDevice.setTextInputItem(textInputItem);

            textInputItem.setOnActionExpandListener(new MenuItem.OnActionExpandListener() {
                @Override
                public boolean onMenuItemActionExpand(final MenuItem item) {
                    hideAllItems(menu);
                    return true;
                }

                @Override
                public boolean onMenuItemActionCollapse(final MenuItem item) {
                    // resets the whole menu, setting `item`'s visibility to true
                    invalidateOptionsMenu();
                    return true;
                }
            });

            final SearchView textInputView = (SearchView) textInputItem.getActionView();
            textInputView.setQueryHint(getResources().getString(R.string.text_input_hint));
            textInputView.setInputType(
                    InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_NORMAL);
            textInputView.setOnQueryTextFocusChangeListener(
                    (v, hasFocus) -> textInputItemFocusJustChanged = true);

        } else {
            // this should be unreachable, but just to be future-proof
            textInputItem.setVisible(false);
        }

        if (appJustOpened) {
            // now everything should have been initialized
            if (!(skillEvaluator.getPrimaryInputDevice() instanceof SpeechInputDevice)
                    || PermissionUtils.checkPermissions(this, RECORD_AUDIO)) {
                // if no voice permission start listening in onActivityResult
                skillEvaluator.getPrimaryInputDevice().tryToGetInput(false);
            }
            appJustOpened = false;
        }
        return true;
    }

    private void hideAllItems(final Menu menu) {
        for (int i = 0; i < menu.size(); ++i) {
            final MenuItem item = menu.getItem(i);
            item.setVisible(false);
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (skillEvaluator != null) {
            skillEvaluator.cancelGettingInput();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        // theme/language changes can cause recreation, so everything will be initialized there
        if (resumingFromSettings && !isRecreating()) {
            // reinitialize everything if resuming from settings
            resumingFromSettings = false;
            initializeSkillEvaluator();
            invalidateOptionsMenu();
            setupVoiceButton();
        }
    }

    @Override
    public boolean onNavigationItemSelected(final MenuItem item) {
        if (item.getItemId() == R.id.action_settings) {
            final Intent intent = new Intent(this, SettingsActivity.class);
            startActivity(intent);
            drawer.closeDrawer(GravityCompat.START);
            resumingFromSettings = true;
        }
        return true;
    }

    @Override
    public void onBackPressed() {
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else if (textInputItem != null && textInputItem.isActionViewExpanded()) {
            invalidateOptionsMenu();
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public void onRequestPermissionsResult(final int requestCode,
                                           @NonNull final String[] permissions,
                                           @NonNull final int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == MICROPHONE_PERMISSION_REQUEST_CODE
                && grantResults.length > 0
                && grantResults[0] == PERMISSION_GRANTED
                && skillEvaluator != null
                && skillEvaluator.getPrimaryInputDevice() instanceof SpeechInputDevice) {
            skillEvaluator.getPrimaryInputDevice().load();
            skillEvaluator.getPrimaryInputDevice().tryToGetInput(false);

        } else if (requestCode == SKILL_PERMISSIONS_REQUEST_CODE
                && skillEvaluator != null) {
            skillEvaluator.onSkillRequestPermissionsResult(grantResults);
        }
        // SETTINGS_PERMISSIONS_REQUEST_CODE results are ignored
    }


    /////////////////////
    // Skill functions //
    /////////////////////

    private void initializeSkillEvaluator() {
        destroySkillEvaluator();

        final InputDevice primaryInputDevice = buildPrimaryInputDevice();

        final ToolbarInputDevice secondaryInputDevice;
        if (primaryInputDevice instanceof ToolbarInputDevice) {
            primaryInputDevice.load();
            secondaryInputDevice = null;
        } else {
            secondaryInputDevice = new ToolbarInputDevice();
            secondaryInputDevice.load();

            if (primaryInputDevice instanceof SpeechInputDevice
                    && !PermissionUtils.checkPermissions(this, RECORD_AUDIO)) {
                ActivityCompat.requestPermissions(this, new String[]{RECORD_AUDIO},
                        MICROPHONE_PERMISSION_REQUEST_CODE);
            } else {
                primaryInputDevice.load(); // load only if permission granted
            }
        }

        final SpeechOutputDevice speechOutputDevice = buildSpeechOutputDevice();
        final GraphicalOutputDevice graphicalOutputDevice = new MainScreenGraphicalDevice(
                findViewById(R.id.outputScrollView), findViewById(R.id.outputLayout));

        SkillHandler.setSkillContextDevices(speechOutputDevice, graphicalOutputDevice);

        skillEvaluator = new SkillEvaluator(
                // Sections language is initialized in BaseActivity.setLocale
                new SkillRanker(SkillHandler.getStandardSkillBatch(),
                        SkillHandler.getFallbackSkill()),
                primaryInputDevice,
                secondaryInputDevice,
                speechOutputDevice,
                graphicalOutputDevice,
                this);
        skillEvaluator.showInitialPanel();
    }

    private InputDevice buildPrimaryInputDevice() {
        final String preference = preferences
                .getString(getString(R.string.pref_key_input_method), "");
        if (preference.equals(getString(R.string.pref_val_input_method_text))) {
            return new ToolbarInputDevice();
        } else { // default
            return new VoskInputDevice(this);
        }
    }

    private SpeechOutputDevice buildSpeechOutputDevice() {
        final String preference = preferences
                .getString(getString(R.string.pref_key_speech_output_method), "");
        if (preference.equals(getString(R.string.pref_val_speech_output_method_nothing))) {
            return new NothingSpeechDevice();
        } else if (preference.equals(getString(R.string.pref_val_speech_output_method_snackbar))) {
            return new SnackbarSpeechDevice(findViewById(android.R.id.content));
        } else if (preference.equals(getString(R.string.pref_val_speech_output_method_toast))) {
            return new ToastSpeechDevice(this);
        } else { // default
            return new AndroidTtsSpeechDevice(this, Sections.getCurrentLocale());
        }
    }

    private void destroySkillEvaluator() {
        if (skillEvaluator != null) {
            skillEvaluator.cleanup();
            skillEvaluator = null;
        }
    }
}

File: app/src/main/java/org/stypox/dicio/MainActivity.kt
package org.stypox.dicio

import android.Manifest.permission
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Bundle
import android.text.InputType
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ProgressBar
import android.widget.ScrollView
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.core.app.ActivityCompat
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.preference.PreferenceManager
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton
import com.google.android.material.navigation.NavigationView
import org.dicio.skill.output.GraphicalOutputDevice
import org.dicio.skill.output.SpeechOutputDevice
import org.stypox.dicio.eval.SkillEvaluator
import org.stypox.dicio.eval.SkillRanker
import org.stypox.dicio.input.InputDevice
import org.stypox.dicio.input.SpeechInputDevice
import org.stypox.dicio.input.ToolbarInputDevice
import org.stypox.dicio.input.VoskInputDevice
import org.stypox.dicio.input.stt_service.SttServiceActivity
import org.stypox.dicio.output.graphical.MainScreenGraphicalDevice
import org.stypox.dicio.output.speech.AndroidTtsSpeechDevice
import org.stypox.dicio.output.speech.NothingSpeechDevice
import org.stypox.dicio.output.speech.SnackbarSpeechDevice
import org.stypox.dicio.output.speech.ToastSpeechDevice
import org.stypox.dicio.settings.SettingsActivity
import org.stypox.dicio.skills.SkillHandler
import org.stypox.dicio.util.BaseActivity
import org.stypox.dicio.util.PermissionUtils

class MainActivity : BaseActivity(), NavigationView.OnNavigationItemSelectedListener {
    private lateinit var preferences: SharedPreferences
    private lateinit var drawer: DrawerLayout
    private lateinit var textInputItem: MenuItem
    private var skillEvaluator: SkillEvaluator? = null
    private var appJustOpened = false
    private var resumingFromSettings = false
    private var textInputItemFocusJustChanged = false

    ////////////////////////
    // Activity lifecycle //
    ////////////////////////
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        preferences = PreferenceManager.getDefaultSharedPreferences(this)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        drawer = findViewById(R.id.drawer_layout)
        val navigationView = findViewById<NavigationView>(R.id.nav_view)

        setSupportActionBar(toolbar)

        val toggle = ActionBarDrawerToggle(
            this, drawer, toolbar, R.string.drawer_open, R.string.drawer_close
        )
        drawer.addDrawerListener(toggle)
        toggle.syncState()

        navigationView.setNavigationItemSelectedListener(this)
        val scrollView = findViewById<ScrollView>(R.id.outputScrollView)
        scrollView.addOnLayoutChangeListener { _, _, top, _, bottom, _, oldTop, _, oldBottom ->
            if (textInputItemFocusJustChanged && (oldBottom != bottom || oldTop != top)) {
                textInputItemFocusJustChanged = false // the keyboard was opened because of menu
                scrollView.postDelayed({
                    scrollView.scrollBy(
                        0,
                        oldBottom - bottom + top - oldTop
                    )
                }, 10)
            }
        }

        appJustOpened = true // determines whether to show initial panel and start listening
        initializeSkillEvaluator()
        setupVoiceButton()
    }

    override fun onDestroy() {
        super.onDestroy()
        destroySkillEvaluator()
        SkillHandler.releaseSkillContext()
    }

    private fun setupVoiceButton() {
        val voiceFab = findViewById<ExtendedFloatingActionButton>(R.id.voiceFab)
        val voiceLoading = findViewById<ProgressBar>(R.id.voiceLoading)
        if (skillEvaluator?.primaryInputDevice is SpeechInputDevice) {
            voiceFab.visibility = View.VISIBLE
            voiceLoading.visibility = View.VISIBLE
            (skillEvaluator?.primaryInputDevice as SpeechInputDevice)
                .setVoiceViews(voiceFab, voiceLoading)
        } else {
            voiceFab.visibility = View.GONE
            voiceLoading.visibility = View.GONE
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        val toolbarInputDevice = (skillEvaluator?.primaryInputDevice as? ToolbarInputDevice)
            ?: skillEvaluator?.secondaryInputDevice
        if (toolbarInputDevice != null) {
            textInputItem = menu.findItem(R.id.action_text_input)
            textInputItem.isVisible = true
            toolbarInputDevice.setTextInputItem(textInputItem)
            textInputItem.setOnActionExpandListener(object : MenuItem.OnActionExpandListener {
                override fun onMenuItemActionExpand(item: MenuItem): Boolean {
                    hideAllItems(menu)
                    return true
                }

                override fun onMenuItemActionCollapse(item: MenuItem): Boolean {
                    // resets the whole menu, setting `item`'s visibility to true
                    invalidateOptionsMenu()
                    return true
                }
            })

            val textInputView = textInputItem.actionView as SearchView
            textInputView.queryHint = resources.getString(R.string.text_input_hint)
            textInputView.inputType =
                InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_NORMAL
            textInputView.setOnQueryTextFocusChangeListener { _, _ ->
                textInputItemFocusJustChanged = true
            }

        } else {
            // this should be unreachable, but just to be future-proof
            textInputItem.isVisible = false
        }

        if (appJustOpened) {
            // now everything should have been initialized
            if (skillEvaluator?.primaryInputDevice !is SpeechInputDevice
                || PermissionUtils.checkPermissions(this, permission.RECORD_AUDIO)
            ) {
                // if no voice permission start listening in onActivityResult
                skillEvaluator?.primaryInputDevice?.tryToGetInput(false)
            }
            appJustOpened = false
        }
        return true
    }

    private fun hideAllItems(menu: Menu) {
        for (i in 0 until menu.size()) {
            val item = menu.getItem(i)
            item.isVisible = false
        }
    }

    override fun onPause() {
        super.onPause()
        skillEvaluator?.cancelGettingInput()
    }

    override fun onResume() {
        super.onResume()

        // theme/language changes can cause recreation, so everything will be initialized there
        if (resumingFromSettings && !isRecreating) {
            // reinitialize everything if resuming from settings
            resumingFromSettings = false
            initializeSkillEvaluator()
            invalidateOptionsMenu()
            setupVoiceButton()
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_settings) {
            val intent = Intent(this, SettingsActivity::class.java)
            startActivity(intent)
            drawer.closeDrawer(GravityCompat.START)
            resumingFromSettings = true
        } else if (item.itemId == R.id.action_stt_service) {
            startActivity(Intent(this, SttServiceActivity::class.java))
            drawer.closeDrawer(GravityCompat.START)
        }
        return true
    }

    override fun onBackPressed() {
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START)
        } else if (textInputItem.isActionViewExpanded) {
            invalidateOptionsMenu()
        } else {
            super.onBackPressed()
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == MICROPHONE_PERMISSION_REQUEST_CODE &&
                grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED &&
                skillEvaluator?.primaryInputDevice is SpeechInputDevice) {
            skillEvaluator?.primaryInputDevice?.load()
            skillEvaluator?.primaryInputDevice?.tryToGetInput(false)
        } else if (requestCode == SKILL_PERMISSIONS_REQUEST_CODE) {
            skillEvaluator?.onSkillRequestPermissionsResult(grantResults)
        }
        // SETTINGS_PERMISSIONS_REQUEST_CODE results are ignored
    }

    /////////////////////
    // Skill functions //
    /////////////////////
    private fun initializeSkillEvaluator() {
        destroySkillEvaluator()
        val primaryInputDevice = buildPrimaryInputDevice()
        val secondaryInputDevice: ToolbarInputDevice?
        if (primaryInputDevice is ToolbarInputDevice) {
            primaryInputDevice.load()
            secondaryInputDevice = null
        } else {
            secondaryInputDevice = ToolbarInputDevice()
            secondaryInputDevice.load()
            if (primaryInputDevice is SpeechInputDevice
                && !PermissionUtils.checkPermissions(this, permission.RECORD_AUDIO)
            ) {
                ActivityCompat.requestPermissions(
                    this, arrayOf(permission.RECORD_AUDIO),
                    MICROPHONE_PERMISSION_REQUEST_CODE
                )
            } else {
                primaryInputDevice.load() // load only if permission granted
            }
        }
        val speechOutputDevice = buildSpeechOutputDevice()
        val graphicalOutputDevice: GraphicalOutputDevice = MainScreenGraphicalDevice(
            findViewById(R.id.outputScrollView), findViewById(R.id.outputLayout)
        )
        SkillHandler.setSkillContextDevices(speechOutputDevice, graphicalOutputDevice)
        skillEvaluator =
            SkillEvaluator( // Sections language is initialized in BaseActivity.setLocale
                SkillRanker(SkillHandler.standardSkillBatch, SkillHandler.fallbackSkill),
                primaryInputDevice,
                secondaryInputDevice,
                speechOutputDevice,
                graphicalOutputDevice,
                this
            )
        skillEvaluator?.showInitialPanel()
    }

    private fun buildPrimaryInputDevice(): InputDevice {
        val preference = preferences
            .getString(getString(R.string.pref_key_input_method), "")
        return if (preference == getString(R.string.pref_val_input_method_text)) {
            ToolbarInputDevice()
        } else { // default
            VoskInputDevice(this)
        }
    }

    private fun buildSpeechOutputDevice(): SpeechOutputDevice {
        val preference = preferences
            .getString(getString(R.string.pref_key_speech_output_method), "")
        return when (preference) {
            getString(R.string.pref_val_speech_output_method_nothing) -> {
                NothingSpeechDevice()
            }
            getString(R.string.pref_val_speech_output_method_snackbar) -> {
                SnackbarSpeechDevice(findViewById(android.R.id.content))
            }
            getString(R.string.pref_val_speech_output_method_toast) -> {
                ToastSpeechDevice(this)
            }
            else -> { // default
                AndroidTtsSpeechDevice(this, Sections.currentLocale)
            }
        }
    }

    private fun destroySkillEvaluator() {
        skillEvaluator?.cleanup()
        skillEvaluator = null
    }

    companion object {
        const val MICROPHONE_PERMISSION_REQUEST_CODE = 13893
        const val SKILL_PERMISSIONS_REQUEST_CODE = 1928430
        const val SETTINGS_PERMISSIONS_REQUEST_CODE = 420938
    }
}